### DISCUSSION TITLE
Wezterm Framerate Concerns

### DISCUSSION THREAD

**Discussion by unphased** [wezterm/wezterm#3664] [Q&A] [answered]
I did some preliminary testing on my M1 Max Macbook Pro by watching the Framerate counter in iStats Menus and using the trackpad to scroll in scrollback, which is typically the least encumbered way to see how fluidly the terminal renders.

It looks like for all of the state of the art macOS terminals these are the framerate numbers that I'm seeing using this very rough testing procedure: 

- Wezterm generally seems to update somewhere around 40fps
- Alacritty seems to go up to 75fps or so
- iTerm2 seems to be able to reach 90fps
- I can just about almost hit 80fps with Kitty

notes on responsiveness feel:

- iTerm2 feels smooth while scrolling but when scroll is released and momentum scroll is happening, the framerate becomes crap... but only once it slows down at the end. A bit hard to describe, and also doesn't really matter. Worth noting, however, that iTerm2 does implement GPU acceleration these days and remains a solid choice of terminal although I have stopped using it for years by now.
- Alacritty/kitty do not really feel much different, both are pretty much satisfactory
- **Wezterm comes in dead last in terms of responsiveness (measurable via the above framerate metric), and i must emphasize here that this is not a "meaningless metric" that I'm talking about. It literally feels like molasses when scrolling around once you compare to these others. For example if you use the trackpad to scroll around in a manpage, it's a real problem.**

Now I will also do a simple evaluation of another basic performance metric, which is the ability of the terminal emulator to sink output. For this I'm just using [this script](https://raw.githubusercontent.com/Dica-Developer/terminal-benchmark/master/test.sh) which essentially dumps a whole ton of chars and times how long it took to do it. The procedure here is just to look at the overall runtime of the script.

- Wezterm comes out on top with 16.297 sec. This was expected since I am under the impression that it is the most optimized terminal today and this reaffirms that notion. 
- Alacritty hauls in a result of 17.179 sec
- Kitty clocks in with a disappointing 32.029 sec
- iTerm2 comes in last with 37.489 sec

Bonus datapoint:

- the script runs in 17.399 sec under tmux locally
- the script runs in 14.911 sec under tmux over ssh on my other machine (an AMD 5950X, not overclocked). This result kinda has me scratching my head! I believe what may be going on is tmux is translating some of these crazy glyphs into simpler ones or something.

Before I digress further... to bring things back full circle now:

These two aspects of performance are both important when it comes to terminal usability. As a high refresh rate aficionado, I would certainly say that the latency and visual consistency benefits of the first notion of performance (i.e. the framerate that the terminal emulator can sustain under a typical low load scenario) is far more important than the raw output sink bitrate. 

I would like to understand if there is a way that we can tweak or configure wezterm to enable superior responsiveness that is on par with these other terms like Alacritty and Kitty! I hope that is possible, because there are a lot of features I'd like to take advantage of with wezterm and turn it into my daily driver. I think the main thing stopping me from doing that right now is the fact that scrolling around inside tmux, vim, less, and scrollback is quite sluggish in feel. I can understand that this is probably not very noticeable on 60hz systems, but I can reassure you that it is very noticeable on 120hz+ displays.

I'd be happy to do more or different testing on my other systems (windows 11 & Ubuntu) but it's probably fair to say that the experience that matters the most to me is the one seen on the M1 Max machine.

**Comment by wez** [comment, marked as answer]
Some settings to experiment with:

* [front_end = "WebGpu"](https://wezfurlong.org/wezterm/config/lua/config/front_end.html) - will more directly use Metal than the OpenGL -> Metal translation used on M1 machines, may yield some more fps.

* These undocumented rendering related cache sizes, particularly the line related ones:
https://github.com/wez/wezterm/blob/d0e9a034406f25af035ecd93f771cc723a2bd704/config/src/config.rs#L748-L757

the default values are:
https://github.com/wez/wezterm/blob/d0e9a034406f25af035ecd93f771cc723a2bd704/config/src/config.rs#L1903-L1921

When rendering a line, BIDI and font shaping are computed and cached and associated with the line, so the initial draw of a screenful of text may take a few milliseconds the first time, but then should be relatively cheap to draw later.  If you are scrolling rapidly through the scrollback then there is a good chance that you will blow those caches most of the time and experience the worst case performance.  Increasing the size of those caches may improve things, but will come at the cost of increased memory utilization.

If what you are experiencing is slowness scrolling through vim or some other "full screen" TUI application then you're exercising both that layer and the output parsing layer, which is additionally subject to this artificial throttle that defaults to 3ms:

https://github.com/wez/wezterm/blob/d0e9a034406f25af035ecd93f771cc723a2bd704/config/src/config.rs#L359-L364

as well as the processing delay for parsing and applying the data to the model.  When a line is updated by output, the cached rendering state for the line is cleared and must be recomputed.

Note that these costs are proportional to the screen area, so larger window sizes cost more than smaller sizes.

You may also wish to try enabling:

https://github.com/wez/wezterm/blob/d0e9a034406f25af035ecd93f771cc723a2bd704/config/src/config.rs#L630-L634

to log a number of metrics/stats that include latency numbers from rendering that may help you reason about the config changes you make as you experiment:

```console
$ wezterm --config periodic_stat_logging=10

STAT                                    current p50 p75  p95
Key(glyph_cache.glyph_cache.hit.rate)   0       283 425  425
Key(glyph_cache.glyph_cache.miss.rate)  0       51  57   57
Key(glyph_cache.image_cache.hit.rate)   5       14  15   26
Key(glyph_cache.image_cache.miss.rate)  0       3   3    3
Key(gui.paint.impl.rate)                5       14  15   25
Key(line_quad_cache.hit.rate)           120     337 361  575
Key(line_quad_cache.miss.rate)          0       50  51   51
Key(line_state_cache.hit.rate)          120     337 361  599
Key(line_to_ele_shape_cache.hit.rate)   5       14  15   59
Key(line_to_ele_shape_cache.miss.rate)  0       4   25   25
Key(mux.pane_output_event.rate)         0       5   9    9
Key(read_from_pane_pty.bytes.rate)      0       475 1351 1351
Key(send_actions_to_mux.rate)           0       5   9    9
Key(shape_cache.hit.rate)               0       39  39   39
Key(shape_cache.miss.rate)              0       7   27   27
Key(window.atlas.allocate.failure.rate) 0       2   2    2
Key(window.atlas.allocate.success.rate) 0       51  67   67

STAT                                             p50      p75      p95
Key(cached_cluster_shape)                        1.39µs   25.47µs  182.27µs
Key(executor.spawn_delay)                        12.10µs  28.16µs  1.27ms
Key(gui.paint.impl)                              921.60µs 1.11ms   2.06ms
Key(paint_pane.lines)                            118.27µs 171.01µs 197.63µs
Key(quad.map)                                    3.15µs   4.67µs   5.31µs
Key(quad_buffer_apply)                           1.90µs   3.10µs   6.11µs
Key(render_screen_line)                          6.50µs   11.84µs  38.40µs
Key(send_actions_to_mux.perform_actions.latency) 7.74µs   15.62µs  1.50ms
Key(shape.harfbuzz)                              4.10µs   7.23µs   19.97µs
Key(window.atlas.allocate.latency)               3.22µs   4.13µs   69.63µs
```

Those same numbers can also be obtained more immediately via lua using the `wezterm.metrics.get_counters()`, `wezterm.metrics.get_latency()`, `wezterm.metrics.get_sizes()`, `wezterm.metrics.get_throughput()` functions via either the debug overlay, or by logging/printing their output in response to triggering a key assignment.

**Reply by unphased** [reply]
Holy crap you're incredible, thanks for all the details!!!
